### DISCUSSION TITLE
0.1.8 on NixOS crashing the system

### DISCUSSION THREAD

**Discussion by djmaze** [niri-wm/niri#618] [General]
Running NixOS 24.05, I was using niri 0.1.6 from NixOS stable before. Today I tried running 0.1.8 from NixOS unstable – from the tty using `exec niri --session`, as I was doing before. niri outputs some logs and then there is just a black screen and the system seems to have been totally halted (not even able to switch to a different tty via keyboard). I need to turn the laptop off. The first time after turning it on again, the screen stays black. I need to turn it off a second time until it gets back to life. So it seems something really severe stuff happening there.

I experienced a few kernel panics before with niri, while something like this never happened with Hyprland.

This is on a Thinkpad X1 Yoga 2nd Gen. I did not yet have time to dig deeper yet – that's why I did not open an issue yet. But is anyone else seeing these kinds of crashes?

**Comment by YaLTeR** [comment]
Black screen usually means that niri failed to initialize the GPU. Though, the rest of the compositor should work, i.e. Mod+Shift+E => Enter should exit it, or, indeed, switching to a different TTY. So that's the weird part that you can't switch to a different TTY.

It can fail to initialize the GPU for example if on NixOS you get a Mesa version mismatch somehow.

Try getting the niri output when it black screens?

**Reply by djmaze** [reply]
Thanks! I could get out of it using Mod+Shift+E. This is an extract of the log:

```
did not find extension DRI_Mesa version 2
did not find extension DRI_IMAGE_DRIVER version 2
failed to bind extensions
did not find extension DRI_Mesa version 2
did not find extension DRI_IMAGE_DRIVER version 2
failed to bind extensions
did not find extension DRI_Mesa version 2
did not find extension DRI_SWRast version 5
failed to bind extensions
2024-08-22T20:56:00.794955Z  WARN niri::backend::tty: error adding device: No such file or directory (os error 2)
2024-08-22T20:56:00.794977Z  INFO niri: listening on Wayland socket: wayland-1
2024-08-22T20:56:00.794983Z  INFO niri: IPC listening on: /run/user/1000/niri.wayland-1.46999.sock
2024-08-22T20:56:02.832390Z  WARN niri::backend::tty: error doing early import: Error::DeviceMissing
2024-08-22T20:56:02.832467Z  WARN niri::backend::tty: error doing early import: Error::DeviceMissing
```

So it seems indeed Mesa-related. Maybe it is a problem as I am mixing in this unstable package into my stable NixOS?

**Reply by djmaze** [reply]
So how can I find out niri's Mesa version?

**Reply by YaLTeR** [reply]
@sodiboo hey, do you know how to check this?

**Reply by sodiboo** [reply]
something like `nix-store -qR $(dirname $(dirname $(readlink $(which niri)))) | rg mesa` should give you the mesa version.

to make it build with the correct mesa version, consider something like `pkgs-unstable.niri.override { inherit (pkgs-stable) mesa; }`.

consider also using niri-flake's `niri-stable` package from its overlay, which should be 0.1.8 and build against the correct version of mesa.

I'd love to elaborate a little more later, but right now I'm sorta busy

**Reply by djmaze** [reply]
With niri 0.1.9, compiling it manually or via niri-flake, I always get a `SIGSEGV` on my system. Any ideas?

**Reply by djmaze** [reply]
Ouch, let me correct that. For niri 0.19:

* When using niri from nixos-unstable, it crashes with SIGSEGV (no helpful log messages).
* When using `niri-stable` via niri-flake, it crashes with SIGSEGV (no helpful log messages).
* When running niri via the included flake, the screen gets black and I can exit niri via shortcut. There are no helpful log messages either.

Starting niri from within a running niri session (opening it in a new window) seems to work with every version.

**Reply by djmaze** [reply]
I found the relevant stacktrace from the coredump for the `niri-stable` crash:

```
#0  0x00007faa692ea7dd in __strlen_avx2 () from /nix/store/3dyw8dzj9ab4m8hv5dpyx7zii8d0w6fi-glibc-2.39-52/lib/libc.so.6
No symbol table info available.
#1  0x00007faa69228d32 in strdup () from /nix/store/3dyw8dzj9ab4m8hv5dpyx7zii8d0w6fi-glibc-2.39-52/lib/libc.so.6
No symbol table info available.
#2  0x00007faa56a55aca in dri2_initialize_drm () from /nix/store/ix0zwzwpxcz5znli5n89344r99jvr558-mesa-24.0.7-drivers/lib/libEGL_mesa.so.0
No symbol table info available.
#3  0x00007faa56a4d900 in dri2_initialize () from /nix/store/ix0zwzwpxcz5znli5n89344r99jvr558-mesa-24.0.7-drivers/lib/libEGL_mesa.so.0
No symbol table info available.
#4  0x00007faa56a3f4a8 in eglInitialize () from /nix/store/ix0zwzwpxcz5znli5n89344r99jvr558-mesa-24.0.7-drivers/lib/libEGL_mesa.so.0
No symbol table info available.
#5  0x000055b81f3286f9 in Initialize () at /build/source/target/x86_64-unknown-linux-gnu/release/build/smithay-823ff3baa7310546/out/egl_bindings.rs:454
No locals.
#6  {closure#1}<gbm::device::Device<smithay::backend::drm::device::fd::DrmDeviceFd>> () at /build/cargo-vendor-dir/smithay-0.3.0/src/backend/egl/display.rs:251
No locals.
#7  wrap_egl_call<u32, smithay::backend::egl::display::{impl#10}::new::{closure_env#1}<gbm::device::Device<smithay::backend::drm::device::fd::DrmDeviceFd>>> () at /build/cargo-vendor-dir/smithay-0.3.0/src/backend/egl/error.rs:164
No locals.
#8  wrap_egl_call_bool<smithay::backend::egl::display::{impl#10}::new::{closure_env#1}<gbm::device::Device<smithay::backend::drm::device::fd::DrmDeviceFd>>> () at /build/cargo-vendor-dir/smithay-0.3.0/src/backend/egl/error.rs:196
No locals.
#9  new<gbm::device::Device<smithay::backend::drm::device::fd::DrmDeviceFd>> () at /build/cargo-vendor-dir/smithay-0.3.0/src/backend/egl/display.rs:250
No locals.
#10 0x000055b81f378012 in device_added () at src/backend/tty.rs:494

[/SNIP]
```

Not sure if this still means the mesa version does not match?

**Reply by sodiboo** [reply]
I believe that's still a mesa version mismatch, yeah. The segfault occurs in [`Initialize`](https://smithay.github.io/smithay/smithay/backend/egl/ffi/egl/fn.Initialize.html) which is called [here within `EGLDisplay::new`](https://github.com/Smithay/smithay/blob/1259a699920c38a5c35e49788313412fde5f323b/src/backend/egl/display.rs#L251) which is in turn called [here by niri](https://github.com/YaLTeR/niri/blob/6a48728ffb1e638839b07f9ab2f06b2adb41dc61/src/backend/tty.rs#L494), and is the entrypoint to the graphics stack.

This isn't how it normally manifests, but still very clearly mesa-related.

When you're using the `niri-flake` package, how exactly are you doing that? `nix run`, adding `niri.packages.x86_64-linux.niri-stable` to your system, or are you applying `niri.overlays.niri` to your packageset and then installing from that? The last approach with the overlays *should* avoid mesa mismatch issues, whereas the other two are not as robust. They usually work on `nixos-unstable` but are almost guaranteed to not work on any stable branch. The overlay actually builds against *your system*, which is why it should always work.

If you're running from the flake in this repo (primarily for development), the story is the same as `nix run` from my flake. That package is also built against `nixos-unstable`. You might have some luck with `nix run github:YaLTeR/niri --override-input nixpkgs github:nixos/nixpkgs/nixos-24.05`.

If none of the above works, try any of the above methods to get a `niri` executable in your path, then `nix run --impure github:nix-community/nixGL -- niri`.

The cleanest way to do this is with the overlay. [It's just one line of configuration](https://github.com/sodiboo/system/blob/5df62b90fe5afc1b4e3d3d3177b41f5f18d06f61/niri.mod.nix#L11-L12).

Note that if you install using the *modules* in my flake as well, the default package is `niri-stable` built against your system (i.e. equivalent to the overlay; although using some trickery to avoid actually applying the overlay unless you explicitly declare so). So if you use those modules to install niri, this "should" be a *zero* line change. But you don't need to use my modules; if you just want my packages, the overlay is the way to go.

**Reply by djmaze** [reply]
I was just using `nix run` when trying `niri-flake`.

Tbh, I am still not very deep into NixOS yet. So not using flakes or overlays for my system yet. Guess I just want to use the `niri-stable` package for now. Probably need to find out about overlays first. Thanks!

**Reply by djmaze** [reply]
Okay, I managed to switch my system configuration to use a flake.nix as well. Now I could add `niri.url = "github:sodiboo/niri-flake";` to the inputs and then add `nixpkgs.overlays = [ inputs.niri.overlays.niri ];` to my configuration.nix, as well as `niri-stable` to `environment.systemPackages`. After rebuilding the system, still the old niri version is symlinked though.

I cannot use your home-manager modules since I do not have my homemanager config switched to flake modules (yet). Is there another way I can replace my system niri with your niri-stable one?

**UPDATE**: I found the newly built package in my nix store and was able to successfully start it. Yeah! Still the question remains on how to properly install it.

**Reply by sodiboo** [reply]
Do you have `pkgs.niri` installed to your home-manager `home.packages`? In that case, this will override the system package in your `PATH`. Try removing niri from your home configuration, and see if the systemwide niri package is now the main one in your `PATH`.

Also note that this is not some kind of hack, it is genuinely better to install niri in your NixOS system configuration because your home manager configuration has no idea what version of mesa your system actually has, whereas the NixOS configuration does.

**Reply by djmaze** [reply]
Ah, no, I guess that was a misunderstanding. I have it in my `environment.systemPackages` in the system `configuration.nix`. And I don't have any niri-related config in my `home.nix`.

Well, nvm, I found the culprit. Trying to install different niri versions, I had used `nix-env` as well, so I had old versions in my profile as well. Reminds me again not to use `nix-env` for installing packages -_-

So, all seems to be working. Thanks for your work!

**Reply by sodiboo** [reply]
> Ah, no, I guess that was a misunderstanding. I have it in my environment.systemPackages in the system configuration.nix. And I don't have any niri-related config in my home.nix.

I don't think i misunderstood. I understood that you had it globally, but since it was not being actually used it was also clearly installed in some user-specific way. I guessed `home.packages` would be the most likely culprit, but yeah `nix-env` would net you the same result. Anything installed as the user will take precedence over the system installation, declarative or not. Glad you figured it out! Have fun with niri :3
